### DISCUSSION TITLE
How to apply correct regex to allow specific undefined references

### DISCUSSION THREAD

**Discussion by markdumay** [remarkjs/remark#1204] [Q&A]
I'm investigating `remark-cli` to lint the Markdown content of a site generated by [Hugo](https://gohugo.io) (a static site generator). So far, I have successfully included `remark-frontmatter` to filter Hugo's front matter. However, I'm struggling to address several `no-undefined-references` warnings. 

The markdown content contains several Hugo `shortcodes` that generate a link. For example, the following line is a relative reference. Hugo replaces the shortcode `relref` with a valid (relative) reference to a local page during build.

```markdown
[title]({{< relref "page" >}})
```

Basically, I would like to allow any reference that contains a shortcode reference to `relref`, `ref`, and `param`, enclosed by `{{<` and `>}}`. I was able to come up with the following configuration and regex pattern (`package.json`):

```json
{
  "name": "remark",
  "version": "0.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "lint:markdown": "remark . --quiet --frail"
  },
  "author": "",
  "license": "ISC",
  "devDependencies": {
    "remark-cli": "^11.0.0",
    "remark-frontmatter": "^4.0.1",
    "remark-lint-list-item-indent": "^3.1.2",
    "remark-lint-no-undefined-references": "^4.2.1",
    "remark-preset-lint-consistent": "^5.1.2",
    "remark-preset-lint-recommended": "^6.1.3"
  },
  "remarkConfig": {
    "plugins": [
      "remark-frontmatter",
      "remark-preset-lint-consistent",
      "remark-preset-lint-recommended",
      [
          "remark-lint-list-item-indent",
          "mixed"
      ],
      [
        "remark-lint-no-undefined-references",
        {
          "allow": [{ "source": "^{{<\\s?(param|ref|relref)\\s+\".*?\"\\s?>}}$" }]
        }
      ]
    ]
  }
}
```

The script seems to work for `[][]` references, but fails on `[]()` references. 

```markdown
<!-- the pattern works when using [][] -->
[ok][{{< relref "a" >}}]
[ok][{{< param "links" >}}]
[ok][{{< ref "links" >}}]
[should fail][{{< xxx "links" >}}]
[should fail][anything]

<!-- the pattern fails when using []() -->
[should not fail]({{< param "fail" >}})
```

I created a [public repository](https://github.com/markdumay/remark) with the above files to illustrate the issue. I'm quite new to remark, so perhaps I made an obvious configuration error. Any help is greatly appreciated!

**Comment by wooorm** [comment]
We parse CommonMark, and optionally GFM or a couple other extensions.
You use a template language that compiles to CommonMark.
We can’t process template languages.

A markdown parser doesn’t see what you have as a link:

```
[title]({{< relref "page" >}})
```

[title]({{< relref "page" >}})

---

See also https://github.com/orgs/remarkjs/discussions/1166, https://github.com/remarkjs/vscode-remark/issues/18, https://github.com/remarkjs/remark-lint/issues/58

**Reply by markdumay** [reply]
Thanks for your quick reply. I realize I might be stretching the application of `remark-lint` beyond its intended scope. Then again, it would be great it there would be a way to strip any unwanted input prior to processing it and creating the AST. This is different from the Liquid discussion, which mentions `markdown -> AST -> markdown`.

For my use case, it would be perfectly fine to filter (strip) any handlebars-like sections (or actually replacing them with other characters/whitespace to preserve line numbers/positions). This discussion mentions an approach https://github.com/mapbox/retext-mapbox-standard/issues/6#issuecomment-189372289, which is pretty close to what I have in mind. Would this be the right starting point, or is there perhaps another approach available?

**Reply by wooorm** [reply]
Yep, if something like that works for you, you can use that!
I don’t know of other alternatives. I also don’t think there’s a hugo/liquid/etc “standard”. That would be nice to have too. If that exists, tooling can be built on it
